File: wpilibj/src/main/java/com/studica/frc/LightTower.java
// Copyright (c) 2026 WPILib contributors.

package com.studica.frc;

import edu.wpi.first.hal.JNIWrapper;

/** Studica Light Tower vendor API backed by five VMX physical DIO outputs. */
public final class LightTower extends JNIWrapper implements AutoCloseable {
  private long handle;
  private int lastStatus;

  /** Creates a tower using physical VMX channels for continuous, R, G, Y, buzzer. */
  public LightTower(int continuous, int red, int green, int yellow, int buzzer) {
    validatePins(continuous, red, green, yellow, buzzer);
    long[] result = create(continuous, red, green, yellow, buzzer);
    if (result == null || result.length < 2) {
      handle = 0;
      lastStatus = -5;
    } else {
      handle = result[0];
      lastStatus = (int) result[1];
    }
  }

  private static void validatePins(int... pins) {
    for (int i = 0; i < pins.length; ++i) {
      if (pins[i] < 0 || pins[i] >= 34) {
        throw new IllegalArgumentException("VMX physical DIO channel out of range");
      }
      for (int j = 0; j < i; ++j) {
        if (pins[i] == pins[j]) {
          throw new IllegalArgumentException("Light Tower outputs must be distinct");
        }
      }
    }
  }

  public synchronized int getLastStatus() {
    return lastStatus;
  }

  public synchronized boolean setRed(boolean enabled) {
    lastStatus = handle == 0 ? -107 : setRed(handle, enabled);
    return lastStatus == 0;
  }

  public synchronized boolean setYellow(boolean enabled) {
    lastStatus = handle == 0 ? -107 : setYellow(handle, enabled);
    return lastStatus == 0;
  }

  public synchronized boolean setGreen(boolean enabled) {
    lastStatus = handle == 0 ? -107 : setGreen(handle, enabled);
    return lastStatus == 0;
  }

  public synchronized boolean setBuzzer(boolean enabled) {
    lastStatus = handle == 0 ? -107 : setBuzzer(handle, enabled);
    return lastStatus == 0;
  }

  public synchronized boolean setSolid() {
    lastStatus = handle == 0 ? -107 : setSolid(handle);
    return lastStatus == 0;
  }

  public synchronized boolean setBlink() {
    lastStatus = handle == 0 ? -107 : setBlink(handle);
    return lastStatus == 0;
  }

  public synchronized boolean off() {
    lastStatus = handle == 0 ? -107 : off(handle);
    return lastStatus == 0;
  }

  public synchronized boolean isAvailable() {
    return handle != 0 && isAvailable(handle);
  }

  @Override
  public synchronized void close() {
    if (handle != 0) {
      destroy(handle);
      handle = 0;
    }
  }

  private static native long[] create(int continuous, int red, int green, int yellow, int buzzer);

  private static native void destroy(long handle);

  private static native int setRed(long handle, boolean enabled);

  private static native int setYellow(long handle, boolean enabled);

  private static native int setGreen(long handle, boolean enabled);

  private static native int setBuzzer(long handle, boolean enabled);

  private static native int setSolid(long handle);

  private static native int setBlink(long handle);

  private static native int off(long handle);

  private static native boolean isAvailable(long handle);
}

File: hal/src/main/native/cpp/LightTower.cpp
// Copyright (c) 2026 WPILib contributors.

#include "studica/LightTower.h"

#include <utility>

namespace studica {

LightTower::LightTower(uint8_t continuous, uint8_t red, uint8_t green,
                       uint8_t yellow, uint8_t buzzer) noexcept {
  m_lastStatus = StudicaLightTower_Create(continuous, red, green, yellow,
                                          buzzer, &m_handle);
}

LightTower::~LightTower() { StudicaLightTower_Destroy(m_handle); }

LightTower::LightTower(LightTower&& other) noexcept
    : m_handle{std::exchange(other.m_handle, 0)},
      m_lastStatus{std::exchange(other.m_lastStatus,
                                 STUDICA_LIGHT_TOWER_NOT_INITIALIZED)} {}

LightTower& LightTower::operator=(LightTower&& other) noexcept {
  if (this == &other) return *this;
  StudicaLightTower_Destroy(m_handle);
  m_handle = std::exchange(other.m_handle, 0);
  m_lastStatus = std::exchange(other.m_lastStatus,
                               STUDICA_LIGHT_TOWER_NOT_INITIALIZED);
  return *this;
}

bool LightTower::SetRed(bool enabled) noexcept {
  m_lastStatus = StudicaLightTower_SetRed(m_handle, enabled);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::SetYellow(bool enabled) noexcept {
  m_lastStatus = StudicaLightTower_SetYellow(m_handle, enabled);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::SetGreen(bool enabled) noexcept {
  m_lastStatus = StudicaLightTower_SetGreen(m_handle, enabled);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::SetBuzzer(bool enabled) noexcept {
  m_lastStatus = StudicaLightTower_SetBuzzer(m_handle, enabled);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::SetSolid() noexcept {
  m_lastStatus = StudicaLightTower_SetSolid(m_handle);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::SetBlink() noexcept {
  m_lastStatus = StudicaLightTower_SetBlink(m_handle);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::Off() noexcept {
  m_lastStatus = StudicaLightTower_Off(m_handle);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK;
}
bool LightTower::IsAvailable() const noexcept {
  uint8_t value = 0;
  m_lastStatus = StudicaLightTower_IsAvailable(m_handle, &value);
  return m_lastStatus == STUDICA_LIGHT_TOWER_OK && value != 0;
}
int32_t LightTower::GetLastStatus() const noexcept { return m_lastStatus; }

}  // namespace studica
